Added lock file when writing combined.json

FDIM opened the pull request:
This PR addresses few things and should not have any breaking changes: 

* Fixes concurrency issues when shardTestFiles flag is used. Currently it might fail randomly as multiple processes attempt to write to the same file at the same time. This is done by using a lock file and postponing write until it's gone.
* Common assets will no longer be copied with each spec execution. Only combined.json and app.js will be updated. 
* Changes default sort function to take into account sessionId - correct me if I am wrong, but this seems like a nice generic solution that works with sharded tests as well? 

Closes #45 , #44

FDIM commented:
@Evilweed I am done with the changes :)

Evilweed commented:
@FDIM if i am correct, sessionID can change during test execution (within the same file). So maybe better option would be to sort **by browser name**, and then by timestamp?

FDIM commented:
@Evilweed Maybe - I am not sure if it can change. What is a bit more strange is that at work, I don't even see sessionId - it's undefined. So I am setting it to process.pid and it seems to work ok. 

```
            var htmlReporter = new HtmlReporter(...);
            var org = { jasmine2MetaDataBuilder: htmlReporter.jasmine2MetaDataBuilder };
            htmlReporter.jasmine2MetaDataBuilder = function () {
                var metadata = org.jasmine2MetaDataBuilder.apply(this, arguments);
                // for some reason session id is not set in current setup
                if (!metadata.sessionId) {
                    metadata.sessionId = process.pid;
                }
                return metadata;
            };
```

So how about using pid?

FDIM commented:
@Evilweed I've added instanceId (process.pid for now) to each spec metadata and reverted default sort order (there is an example for the one with instanceId).

Additionally I've changed file lock to be a directory instead, as I still had the concurrency error few times at work. 

It would be great if you could merge this or let me know what else I should do before you can merge :)

Evilweed commented:
@FDIM If Your sorting function (the one that takes into account the process.id) makes HTML report look properly when sharding is set to **False** as well as when it's set to **True** then i would recommend making Your sorting function default.

Can You make Your implementation default, and update doc file?

Evilweed commented:
@FDIM Merged. I'll do things i mentioned above.
